assets/js/phoenix_live_view/hooks.js

```diff
@@ -0,0 +1,34 @@
+import {
+  PHX_PREFLIGHTED_REFS,
+  PHX_UPLOAD_REF
+} from "phoenix_live_view/constants"
+
+import LiveUploader from "phoenix_live_view/live_uploader"
+
+let Hooks = {
+  LiveFileUpload: {
+    preflightedRefs(){ return this.el.getAttribute(PHX_PREFLIGHTED_REFS) },
+
+    mounted(){ this.preflightedWas = this.preflightedRefs() },
+
+    updated(){
+      let newPreflights = this.preflightedRefs()
+      if(this.preflightedWas !== newPreflights){
+        this.preflightedWas = newPreflights
+        if(newPreflights === ""){
+          this.__view.cancelSubmit(this.el.form)
+        }
+      }
+    }
+  },
+
+  LiveImgPreview: {
+    mounted(){
+      this.ref = this.el.getAttribute("data-phx-entry-ref")
```

Comment:
This PR is a great resource for learning how to separate concerns. Thank you.

I'm curious, is this attribute not constantised because it's the only place it's used?

Reply:
This particular LOC wasn't changed when extracted into the class-per-file style. After this is merged and vetted we can probably take a look at some refactors.